<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8' />
    <meta http-equiv="X-UA-Compatible" content="chrome=1" />
    <meta name="description" content="annie : ANNotation Information Extractor" />

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>annie</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/genomeannotation/Annie">View on GitHub</a>

          <h1 id="project_title">annie</h1>
          <h2 id="project_tagline">ANNotation Information Extractor</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/genomeannotation/Annie/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/genomeannotation/Annie/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <p>This page is under development. Thank you for your patience!</p>

<h1>
<a name="introduction" class="anchor" href="#introduction"><span class="octicon octicon-link"></span></a>Introduction</h1>

<h3>
<a name="license" class="anchor" href="#license"><span class="octicon octicon-link"></span></a>License</h3>

<p>annie is released under the <a href="http://opensource.org/licenses/MIT">MIT License</a></p>

<h3>
<a name="purpose" class="anchor" href="#purpose"><span class="octicon octicon-link"></span></a>Purpose</h3>

<p>The purpose of annie is relatively simple. annie takes your input files (blastout, fasta, gff, etc), does the appropriate cross-referencing for you, and spits out a file that contains the appropriate annotations written in a 3-column table format.  Why would I want to create a 3 column table in this format?  To add functional annotations to my genome using the <a href="http://genomeannotation.github.io/GAG/"> Genome Annotation Generator (GAG)</a> of course!</p>

<h3>
<a name="installation" class="anchor" href="#installation"><span class="octicon octicon-link"></span></a>Installation</h3>

<p>annie is written in the beautiful language of <a href="https://www.python.org/download/releases/2.7.6/">Python 2</a> . If you have Python 2 (not 3!) installed on your computer, you can download the zipped source code link at the top of the page and extract it. That's it!</p>

<h3>
<a name="citing-annie" class="anchor" href="#citing-annie"><span class="octicon octicon-link"></span></a>Citing annie</h3>

<p>If you'd like to flatter us, please cite us as follows:</p>

<blockquote>
<p>Tate, R., Hall, B., DeRego, T., &amp; Geib, S. (2014). annie: the ANNotation Information Extractor (Version 1.0) [Software]. Available from <a href="http://genomeannotation.github.io/Annie">http://genomeannotation.github.io/Annie</a>.</p>
</blockquote>

<h1>
<a name="usage" class="anchor" href="#usage"><span class="octicon octicon-link"></span></a>Usage</h1>

<h3>
<a name="getting-started" class="anchor" href="#getting-started"><span class="octicon octicon-link"></span></a>Getting Started</h3>

<p>First, open up your command prompt/terminal traverse your directories until your reach the annie directory (wherever you decided to extract it on your computer). Once there, you will want to run a command in the following form:</p>

<p><code>python annie.py &lt;case&gt; &lt;input_file1&gt; [&lt;input_file2&gt; ...] &lt;output_file&gt;</code></p>

<p>Currently, Annie works with two cases: Results from InterProScan (ipr) and BLAST searching the UniProt SwissProt database (sprot). More information about the inputFile and outputFile fields are outlined below.</p>

<p>If you ever get lost and want to know what the valid inputs for annie are just type:
<code>python annie.py help</code>
and help will be on the way shortly.</p>

<h3>
<a name="ipr" class="anchor" href="#ipr"><span class="octicon octicon-link"></span></a>InterProScan (ipr)</h3>

<p>For InterProScan (ipr), the command has the following form:</p>

<p><code>python annie.py ipr &lt;ipr_input_file&gt; &lt;output_file&gt;</code></p>

<p>The ipr input file is the tab separate (.tsv) output from InterProScan, make sure you have this format (not xml).   An example of what an actual command might look like would be:</p>

<p><code>python annie.py ipr my_ipr_file.tsv my_super_secret_output_file.txt</code></p>

<p>The output file name can be whatever you want it to be, that's your chance to be creative! Another thing to note is that the file extensions don't matter for either the input and output files. For example, your ipr file could have an extension <code>.animu</code> and everything would work just fine as long as it actually is the tab-separated output from InterProScan.</p>

<p>The output file is a table of annotations. For each row in the ipr file, it contains the mrna id and the dbxref. It also contains the GO and IPR values if they exist. For each row, we create up to 3 annotations where the feature id is the mrna id, and the keys are "dbxref", "IPR" and "GO" with their respective values. Here is a sample input/output pair:</p>

<p>Input:</p>

<p><code>m.98281    c95b0824ccd627403aa63f9e474649cc    7571    Pfam    PF00041 Fibronectin type III domain 3729    3812    4.7E-14 T   04-04-2014  IPR003961   Fibronectin, type III   GO:0005515</code> </p>

<p><code>m.98281    c95b0824ccd627403aa63f9e474649cc    7571    Pfam    PF00041 Fibronectin type III domain 6484    6567    1.8E-12 T   04-04-2014  IPR003961   Fibronectin, type III   GO:0005515</code></p>

<p><code>m.42655    de17ff06d901d22dacc3f5c91510f33f    288 Pfam    PF12171 Zinc-finger double-stranded RNA-binding 45  69  1.3E-5  T   05-04-2014  IPR022755   Zinc finger, double-stranded RNA binding</code>   </p>

<p><code>m.82734    ae22a6fcc80b2ac982378f16ac022b3d    144 Pfam    PF14846 Domain of unknown function (DUF4485)    8   89  4.8E-17 T   05-04-2014  IPR027831   Domain of unknown function DUF4485</code></p>

<p>Output:</p>

<p><code>m.42655    Dbxref  PFAM:PF12171</code></p>

<p><code>m.42655    InterPro    IPR022755</code></p>

<p><code>m.82734    Dbxref  PFAM:PF14846</code></p>

<p><code>m.82734    InterPro    IPR027831</code></p>

<p><code>m.98281    Dbxref  PFAM:PF00041</code></p>

<p><code>m.98281    GO  GO:0005515</code></p>

<p><code>m.98281    InterPro    IPR003961</code></p>

<p>A few notes on the sample input/output:</p>

<ul>
<li>The GO annotation doesn't appear for every mrna because not every mrna has one.</li>
<li>Notice that the mrna, dbxref, GO, and IPR values for the first two entries in the input are the same so they result in the same annotations. Knowing that nobody has time for duplicate annotations, annie automatically removes them for you.</li>
<li>The output has the annotations <em>ordered</em>. In other words, it first sorts by the feature id (the mrna id), then by the key (e.g. "Dbxref", "GO", "InterPro"), then by the value.</li>
</ul><p>Annie makes a few assumptions about your input files so if something doesn't work, check to see if your file meets these assumptions:</p>

<ul>
<li>values in each row are tab-separated</li>
<li>the first column is the mrna id</li>
<li>the 4th and 5th columns correspond to your dbxref</li>
<li>the 12th column is the IPR column if the 12th column exists</li>
<li>the 14th column is the GO column if the 14th column exists</li>
</ul><h3>
<a name="sprot" class="anchor" href="#sprot"><span class="octicon octicon-link"></span></a>SwissProt (sprot)</h3>

<p>For gene names and protein functions from SwissProt (sprot), the command has the following form, it's a bit longer compared to ipr:</p>

<p><code>python annie.py sprot &lt;blastout_file_name&gt; &lt;gff_file_name&gt; &lt;fasta_file_name&gt; &lt;output_file_name&gt;</code></p>

<p>So essentially, we take three input files (blast, gff, and fasta).  To be clear, the blastout_file should be the blast result file from your search against the SwissProt database.  This output should be in tabular format (blastall -m 8 or blast+ -outfmt 6).  XML or plain text format cannot be read (sorry).  The gff_file is the GFF3 file containing your genome (or transcriptome) annotation.  This is so Annie knows what the parent gene is of each mRNA ID (see below for description).  The fasta_file is the SwissProt fasta file you performed blast search against (the database).  That big giant file you downloaded <a href="http://www.uniprot.org/downloads">here</a>.  Use the UniProtKB/SwissProt Database, those are manually currated genes.  You can send us a nasty email or spread rumors about Annie if you think this is not the best database to use (of course nothing is perfect).     An example of what an actual command might look like would be:</p>

<p><code>python annie.py sprot my_blast_file.blast my_gff_file.gff my_fasta_file.fasta my_results.txt</code></p>

<p>Once again, the output file can be named anything you want. Another thing to note is that the file extensions don't matter for either the input and output files. For example, your gff file could have an extension <code>.velociraptor</code> and everything would work just fine.</p>

<p>The output file is a table of annotations just like the ipr case. The sprot case is a bit more complicated so we'll go through the details before moving on to the sample input/output. First of all, the main goal is to write annotations for all the mrna's in the blastout file. Secondly, for each of those mrna's, we want to write annotations for their corresponding parent gene. </p>

<p>The sprot case returns two types of annotations. The first is a product annotation in the form <code>&lt;mrna_id&gt; product &lt;product&gt;</code>. We use the blast file to get the dbxref for the associated mrna and then we use the fasta file to take that dbxref and get the corresponding product. Here is a diagram of what goes on:</p>

<p><code>mrna_id ---blast_file---&gt; dbxref ---fasta_file---&gt; product</code></p>

<p>Next, we have the name annotation. The name annotation has the form <code>&lt;parent_gene_id&gt; name &lt;parent_gene_name&gt;</code>. First, for each mrna in the blast results, we look it up in the gff file to get the corresponding parent gene id. That is how we obtain the <code>&lt;parent_gene_id&gt;</code> portion of the annotation. The process to obtain the <code>&lt;parent_gene_name&gt;</code> is similar to getting the product for the product annotation. First, we use the blast file to get the associated dbxref from the mrna and then we use the fasta file to take that dbxref ref and give us the gene name. Here is a diagram of how the name annotation is made:</p>

<p><code>mrna_id ---gff_file---&gt; parent_gene_id</code></p>

<p><code>mrna_id ---blast_file---&gt; dbxref ---fasta_file---&gt; parent_gene_name</code></p>

<p>Now that we understand what this case is doing, let's take a look at some sample input/output:</p>

<h5>
<a name="input" class="anchor" href="#input"><span class="octicon octicon-link"></span></a>Input</h5>

<h6>
<a name="blast-file" class="anchor" href="#blast-file"><span class="octicon octicon-link"></span></a>blast file</h6>

<p><code>m.4830 sp|Q5AZY1|MRH4_EMENI    32.65   49  33  0   114 162 500 548 0.56    34.3</code></p>

<p><code>m.4831 sp|Q9TTC1|POL_KORV  32.26   155 102 3   23  174 807 961 1e-16   81.6</code></p>

<p><code>m.4837 sp|P05892|GAG_SIVVT 45.24   42  21  2   35  75  394 434 0.012   38.1</code></p>

<p><code>m.4838 sp|Q9UGP4|LIMD1_HUMAN   29.58   71  42  3   32  96  369 437 0.88    31.6</code></p>

<h6>
<a name="gff-file" class="anchor" href="#gff-file"><span class="octicon octicon-link"></span></a>gff file</h6>

<p><code>comp9975_c0_seq1   .   gene    25  603 .   +   .   ID=g.4830;Name=ORF%20g.4830%20m.4830%20type%3Acomplete%20len%3A193%20%28%2B%29</code></p>

<p><code>comp9975_c0_seq1   .   mRNA    25  603 .   +   .   ID=m.4830;Parent=g.4830</code></p>

<p><code>comp9975_c0_seq1   .   CDS 25  603 .   +   .   ID=cds.m.4830;Parent=m.4830</code></p>

<p><code>comp9975_c1_seq1   .   gene    3   533 .   +   .   ID=g.4831;Name=ORF%20g.4831%20m.4831%20type%3Ainternal%20len%3A176%20%28%2B%29</code></p>

<p><code>comp9975_c1_seq1   .   mRNA    3   533 .   +   .   ID=m.4831;Parent=g.4831</code></p>

<p><code>comp9975_c1_seq1   .   CDS 3   533 .   +   .   ID=cds.m.4831;Parent=m.4831</code></p>

<p><code>comp9982_c0_seq1   .   gene    1   441 .   -   .   ID=g.4837;Name=ORF%20g.4837%20m.4837%20type%3Ainternal%20len%3A147%20%28-%29</code></p>

<p><code>comp9982_c0_seq1   .   mRNA    1   441 .   -   .   ID=m.4837;Parent=g.4837</code></p>

<p><code>comp9982_c0_seq1   .   CDS 1   441 .   -   .   ID=cds.m.4837;Parent=m.4837</code></p>

<p><code>comp9983_c0_seq1   .   gene    1746    2060    .   +   .   ID=g.4838;Name=ORF%20g.4838%20m.4838%20type%3A3prime_partial%20len%3A104%20%28%2B%29</code></p>

<p><code>comp9983_c0_seq1   .   mRNA    1746    2060    .   +   .   ID=m.4838;Parent=g.4838</code></p>

<p><code>comp9983_c0_seq1   .   CDS 1746    2060    .   +   .   ID=cds.m.4838;Parent=m.4838</code></p>

<h6>
<a name="fasta-file" class="anchor" href="#fasta-file"><span class="octicon octicon-link"></span></a>fasta file</h6>

<p>(we removed the sequences for readability of this example)</p>

<p><code>&gt;sp|Q5AZY1|MRH4_EMENI ATP-dependent RNA helicase mrh4, mitochondrial OS=Emericella nidulans (strain FGSC A4 / ATCC 38163 / CBS 112.46 / NRRL 194 / M139) GN=mrh4 PE=3 SV=1</code></p>

<p><code>&gt;sp|Q9TTC1|POL_KORV Pro-Pol polyprotein OS=Koala retrovirus GN=pro-pol PE=3 SV=1</code></p>

<p><code>&gt;sp|P05892|GAG_SIVVT Gag polyprotein OS=Simian immunodeficiency virus agm.vervet (isolate AGM TYO-1) GN=gag PE=3 SV=1</code></p>

<p><code>&gt;sp|Q9UGP4|LIMD1_HUMAN LIM domain-containing protein 1 OS=Homo sapiens PE=1 SV=1</code></p>

<h5>
<a name="output" class="anchor" href="#output"><span class="octicon octicon-link"></span></a>Output</h5>

<p><code>g.4838 name    LIMD1</code></p>

<p><code>m.4838 product LIM domain-containing protein 1</code></p>

<p><code>g.4830 name    mrh4</code></p>

<p><code>m.4830 product ATP-dependent RNA helicase mrh4, mitochondrial</code></p>

<p><code>g.4831 name    pro-pol</code></p>

<p><code>m.4831 product Pro-Pol polyprotein</code></p>

<p><code>g.4837 name    gag</code></p>

<p><code>m.4837 product Gag polyprotein</code></p>

<p>A few notes on the sample input/output:</p>

<ul>
<li>Not every gene id has an associated gene name. In that case, we use part of the dbxref to name it (the part after the second '|' and from that part, the part before the first '_').</li>
<li>Although not seen here, if two name annotations have the same value (i.e. the same parent gene name), we index them by adding a "_0" in front of the first one, a "_1" in front of the second one, etc.</li>
<li>Although not seen here, if annie fails to cross-reference something, it'll skip doing that annotation and gently notify you in the command prompt/terminal.</li>
</ul><p>Annie makes a few assumptions about your input files so if something doesn't work, check to see if your file meets these assumptions:</p>

<ul>
<li>in the blast file, every row has tab-separated values with the mrna id in the first row and the dbxref in the second row</li>
<li>in the gff file, every row has tab separated values</li>
<li>in the gff file, the third column has the feature type ("mRNA", etc). We also assume the third column, if it is for mrna, has exactly the value "mRNA" with no surrounding whitespace or changes in captilization.</li>
<li>in the gff file, the 9th column should have the mrna id and parent gene id in exactly the following form: <code>ID=&lt;mrna_id&gt;;Parent=&lt;parent_gene_id&gt;</code> so an example would ne <code>ID=m.123;Parent=g.123</code>
</li>
<li>in the fasta file, if the line is a header line, the first character is '&gt;'</li>
<li>in the fasta file, the dbxref is immediately the first thing to appear after the '&gt;' with no whitespace inbetween the '&gt;' and the dbxref</li>
<li>in the fasta file, the dbxref has no whitespace contained inside</li>
<li>in the fasta file, the product is everything between the dbxref and the string "OS=" which we assume exists</li>
<li>in the fasta file, if the gene name exists, it contains no whitespace within</li>
<li>in the fasta file, if the gene name exists, it is right before the string "PE=" which we assume exists</li>
<li>in the fasta file, we assume the gene name is in the following form: <code>GN=&lt;gene_name&gt;</code>. An example would be <code>GN=gag</code>
</li>
</ul><h1>
<a name="configuration" class="anchor" href="#configuration"><span class="octicon octicon-link"></span></a>Configuration</h1>

<p>Currently, annie only has one type of configuration: the whitelist. More information below.</p>

<h3>
<a name="dbxref-whitelist" class="anchor" href="#dbxref-whitelist"><span class="octicon octicon-link"></span></a>Dbxref Whitelist</h3>

<p>The whitelist is used for the ipr case. If the database type (e.g. 'Pfam') is not in the whitelist, then the annotation won't be written. In other words, annotations are only written for <em>valid</em> databases contained in the whitelist file. annie will convert both the whitelist entry and the ipr file entry to lowercase and trim surrounding whitespace before comparing them against each other, i.e., annie doesn't care about case-sensitivity or whitespace when it comes to the databases names in the ipr file and the whitelist.</p>

<p>To edit the whitelist, go to the Annie directory, then go to the config subdirectory. As of the latest version of annie, there should only be the whitelist file in there. Feel free to delete entries as desired or to add new entries on a new line. Or delete <em>all</em> the entries! Up to you; go wild my friend.</p>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">annie maintained by <a href="https://github.com/genomeannotation">genomeannotation</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
